src/syntaxtree.c

```diff
@@ -0,0 +1,898 @@
+{
+    Obj first;
+    Obj second;
+    Obj last;
+
+    if (SIZE_EXPR(expr) == 2 * sizeof(Expr)) {
+        first = SyntaxTreeCompiler(ADDR_EXPR(expr)[0]);
+        last = SyntaxTreeCompiler(ADDR_EXPR(expr)[1]);
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("last"), last);
+    }
+    else {
+        first = SyntaxTreeCompiler(ADDR_EXPR(expr)[0]);
+        second = SyntaxTreeCompiler(ADDR_EXPR(expr)[1]);
+        last = SyntaxTreeCompiler(ADDR_EXPR(expr)[2]);
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("second"), second);
+        AssPRec(result, RNamName("last"), last);
+    }
+
+    return result;
+}
+
+Obj SyntaxTreeStringExpr(Obj result, Expr expr)
+{
+    Obj string;
+
+    C_NEW_STRING(string, SIZE_EXPR(expr) - 1 - sizeof(UInt),
+                 sizeof(UInt) + (Char *)ADDR_EXPR(expr));
+
+    AssPRec(result, RNamName("string"), string);
+
+    return result;
+}
+
+/* TODO: Maybe return a record or a list with key/value pairs? */
+Obj SyntaxTreeRecExpr(Obj result, Expr expr)
+{
+    Obj  key;
+    Obj  val;
+    Obj  list;
+    Obj  subrec;
+    Expr tmp;
+    Int  i, len;
+
+    len = SIZE_EXPR(expr) / (2 * sizeof(Expr));
+    list = NEW_PLIST(T_PLIST, len);
+    SET_LEN_PLIST(list, len);
+
+    for (i = 1; i <= len; i++) {
+        // Check whether this can happen and *why*
+        tmp = ADDR_EXPR(expr)[2 * i - 2];
+        if (tmp == 0) {
+            continue;
+        }
+        else {
+            subrec = NEW_PREC(2);
+            SET_ELM_PLIST(list, i, subrec);
+            CHANGED_BAG(list);
+
+            if (IS_INTEXPR(tmp)) {
+                key = NAME_RNAM((UInt)INT_INTEXPR(tmp));
+            }
+            else {
+                key = SyntaxTreeCompiler(tmp);
+            }
+            AssPRec(subrec, RNamName("key"), key);
+
+            tmp = ADDR_EXPR(expr)[2 * i - 1];
+            val = SyntaxTreeCompiler(tmp);
+            AssPRec(subrec, RNamName("value"), val);
+        }
+    }
+    AssPRec(result, RNamName("keyvalue"), list);
+
+    return result;
+}
+
+/* TODO: Slightly ugly hack */
+extern Obj EAGER_FLOAT_LITERAL_CACHE;
+Obj SyntaxTreeFloatEager(Obj result, Expr expr)
+{
+    UInt ix = READ_EXPR(expr, 0);
+    AssPRec(result, RNamName("value"), ELM_LIST(EAGER_FLOAT_LITERAL_CACHE, ix));
+    return result;
+}
+
+Obj SyntaxTreeFloatLazy(Obj result, Expr expr)
+{
+    UInt len;
+    Obj string;
+
+    len = READ_EXPR(expr, 0);
+    string = NEW_STRING(len);
+    memcpy(CHARS_STRING(string),
+           (const char *)CONST_ADDR_EXPR(expr) + 2 * sizeof(UInt), len);
+    AssPRec(result, RNamName("value"), string);
+    return result;
+}
+
+Obj SyntaxTreeLVar(Expr expr)
+{
+    LVar lvar;
+
+    if (IS_REFLVAR(expr)) {
+        lvar = LVAR_REFLVAR(expr);
+    } else {
+        lvar = (LVar)(ADDR_EXPR(expr)[0]);
+    }
+
+    return INTOBJ_INT(lvar);
+}
+
+Obj SyntaxTreeHVar(Expr expr)
+{
+    HVar hvar;
+
+    hvar = (HVar)(expr);
```

Comment:
Huh -- shouldn't this rather be `hvar = *ADDR_EXPR(expr);`?

EDIT: Oh, I get it now: this function is only used as an entry for an `argcomp` field, which already performed the ADDR_EXPR for us.

I think this would be much clearer if the function names reflected it. E.g. `SyntaxTreeArgCompHVar`, or just  `static ArgCompHVar` or so.

And similarly, the "compiler" functions could have a separate name. That would make a lot of things clearer to me.